File: Backend/date_utils.py
"""
Date utility functions for consistent DD/MM/YYYY format throughout the backend
"""

from datetime import datetime, date
import re
import logging

logger = logging.getLogger(__name__)

def parse_ddmmyyyy_to_date(ddmmyyyy_string):
    """
    Parse DD/MM/YYYY format string to Python date object
    
    Args:
        ddmmyyyy_string (str): Date string in DD/MM/YYYY format
        
    Returns:
        date: Python date object or None if invalid
    """
    if not ddmmyyyy_string or not isinstance(ddmmyyyy_string, str):
        return None
    
    try:
        # Handle DD/MM/YYYY format
        if re.match(r'^\d{2}/\d{2}/\d{4}$', ddmmyyyy_string):
            day, month, year = ddmmyyyy_string.split('/')
            return date(int(year), int(month), int(day))
        
        # Handle other formats as fallback
        # Try to parse as ISO format (YYYY-MM-DD)
        if re.match(r'^\d{4}-\d{2}-\d{2}$', ddmmyyyy_string):
            return datetime.strptime(ddmmyyyy_string, '%Y-%m-%d').date()
        
        # Try to parse as datetime and extract date
        parsed_date = datetime.fromisoformat(ddmmyyyy_string.replace('Z', '+00:00'))
        return parsed_date.date()
        
    except (ValueError, TypeError) as e:
        logger.warning(f"Failed to parse date '{ddmmyyyy_string}': {e}")
        return None

def format_date_to_ddmmyyyy(date_obj):
    """
    Format Python date/datetime object to DD/MM/YYYY string
    
    Args:
        date_obj: Python date or datetime object
        
    Returns:
        str: Date string in DD/MM/YYYY format
    """
    if not date_obj:
        return ''
    
    try:
        if isinstance(date_obj, datetime):
            date_obj = date_obj.date()
        
        if isinstance(date_obj, date):
            return date_obj.strftime('%d/%m/%Y')
        
        return str(date_obj)
        
    except (AttributeError, TypeError) as e:
        logger.warning(f"Failed to format date '{date_obj}': {e}")
        return str(date_obj)

def parse_ddmmyyyy_to_datetime(ddmmyyyy_string, time_string='00:00'):
    """
    Parse DD/MM/YYYY format string with time to Python datetime object
    
    Args:
        ddmmyyyy_string (str): Date string in DD/MM/YYYY format
        time_string (str): Time string in HH:MM format
        
    Returns:
        datetime: Python datetime object or None if invalid
    """
    date_part = parse_ddmmyyyy_to_date(ddmmyyyy_string)
    if not date_part:
        return None
    
    try:
        # Parse time if provided
        if time_string and isinstance(time_string, str):
            # Handle HH:MM:SS format
            if ':' in time_string:
                time_parts = time_string.split(':')
                hour = int(time_parts[0])
                minute = int(time_parts[1])
                second = int(time_parts[2]) if len(time_parts) > 2 else 0
            else:
                hour = minute = second = 0
        else:
            hour = minute = second = 0
        
        return datetime.combine(date_part, datetime.min.time().replace(
            hour=hour, minute=minute, second=second
        ))
        
    except (ValueError, TypeError) as e:
        logger.warning(f"Failed to parse datetime '{ddmmyyyy_string} {time_string}': {e}")
        return None

def convert_html_date_to_ddmmyyyy(html_date_string):
    """
    Convert YYYY-MM-DD format (from HTML date input) to DD/MM/YYYY
    
    Args:
        html_date_string (str): Date string in YYYY-MM-DD format
        
    Returns:
        str: Date string in DD/MM/YYYY format
    """
    if not html_date_string or not isinstance(html_date_string, str):
        return ''
    
    try:
        if re.match(r'^\d{4}-\d{2}-\d{2}$', html_date_string):
            date_obj = datetime.strptime(html_date_string, '%Y-%m-%d').date()
            return date_obj.strftime('%d/%m/%Y')
        
        return html_date_string
        
    except (ValueError, TypeError) as e:
        logger.warning(f"Failed to convert HTML date '{html_date_string}': {e}")
        return html_date_string

def convert_ddmmyyyy_to_html_date(ddmmyyyy_string):
    """
    Convert DD/MM/YYYY format to YYYY-MM-DD format (for HTML date input)
    
    Args:
        ddmmyyyy_string (str): Date string in DD/MM/YYYY format
        
    Returns:
        str: Date string in YYYY-MM-DD format
    """
    if not ddmmyyyy_string or not isinstance(ddmmyyyy_string, str):
        return ''
    
    try:
        if re.match(r'^\d{2}/\d{2}/\d{4}$', ddmmyyyy_string):
            day, month, year = ddmmyyyy_string.split('/')
            return f"{year}-{month.zfill(2)}-{day.zfill(2)}"
        
        return ddmmyyyy_string
        
    except (ValueError, TypeError) as e:
        logger.warning(f"Failed to convert DD/MM/YYYY date '{ddmmyyyy_string}': {e}")
        return ddmmyyyy_string

def is_valid_ddmmyyyy(date_string):
    """
    Validate if a string is a valid DD/MM/YYYY date
    
    Args:
        date_string (str): Date string to validate
        
    Returns:
        bool: True if valid DD/MM/YYYY format
    """
    if not date_string or not isinstance(date_string, str):
        return False
    
    try:
        if not re.match(r'^\d{2}/\d{2}/\d{4}$', date_string):
            return False
        
        day, month, year = date_string.split('/')
        day, month, year = int(day), int(month), int(year)
        
        # Check if it's a valid date
        test_date = date(year, month, day)
        return test_date.day == day and test_date.month == month and test_date.year == year
        
    except (ValueError, TypeError):
        return False

def get_current_date_ddmmyyyy():
    """
    Get current date in DD/MM/YYYY format
    
    Returns:
        str: Current date in DD/MM/YYYY format
    """
    return datetime.now().strftime('%d/%m/%Y')

def get_current_datetime_ddmmyyyy():
    """
    Get current datetime in DD/MM/YYYY HH:MM format
    
    Returns:
        str: Current datetime in DD/MM/YYYY HH:MM format
    """
    return datetime.now().strftime('%d/%m/%Y %H:%M')

def format_datetime_for_display(datetime_obj):
    """
    Format datetime object for display in DD/MM/YYYY HH:MM format
    
    Args:
        datetime_obj: Python datetime object
        
    Returns:
        str: Formatted datetime string
    """
    if not datetime_obj:
        return ''
    
    try:
        if isinstance(datetime_obj, str):
            # Try to parse the string first
            datetime_obj = datetime.fromisoformat(datetime_obj.replace('Z', '+00:00'))
        
        if isinstance(datetime_obj, datetime):
            return datetime_obj.strftime('%d/%m/%Y %H:%M')
        elif isinstance(datetime_obj, date):
            return datetime_obj.strftime('%d/%m/%Y')
        
        return str(datetime_obj)
        
    except (ValueError, TypeError, AttributeError) as e:
        logger.warning(f"Failed to format datetime '{datetime_obj}': {e}")
        return str(datetime_obj)

File: hrmshiring-main/src/utils/dateUtils.js
/**
 * Date utility functions for consistent DD/MM/YYYY format throughout the application
 */

/**
 * Format a date string or Date object to DD/MM/YYYY format for display
 * @param {string|Date} dateInput - Date string or Date object
 * @returns {string} - Formatted date in DD/MM/YYYY format
 */
export const formatDateForDisplay = (dateInput) => {
  if (!dateInput) return 'Not specified';
  
  try {
    const date = new Date(dateInput);
    if (isNaN(date.getTime())) {
      return dateInput.toString();
    }
    
    const day = String(date.getDate()).padStart(2, '0');
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const year = date.getFullYear();
    
    return `${day}/${month}/${year}`;
  } catch (error) {
    console.error('Error formatting date for display:', error);
    return dateInput.toString();
  }
};

/**
 * Format a date string or Date object to DD/MM/YYYY format with time
 * @param {string|Date} dateInput - Date string or Date object
 * @param {string} timeInput - Time string (optional)
 * @returns {string} - Formatted date and time in DD/MM/YYYY HH:MM format
 */
export const formatDateTimeForDisplay = (dateInput, timeInput = null) => {
  if (!dateInput) return 'Not specified';
  
  try {
    const date = new Date(dateInput);
    if (isNaN(date.getTime())) {
      return dateInput.toString();
    }
    
    const day = String(date.getDate()).padStart(2, '0');
    const month = String(date.getMonth() + 1).padStart(2, '0');
    const year = date.getFullYear();
    
    let formattedDate = `${day}/${month}/${year}`;
    
    if (timeInput) {
      // Format time to HH:MM
      let timeStr = timeInput;
      if (timeStr.includes(':')) {
        const [hours, minutes] = timeStr.split(':');
        timeStr = `${String(hours).padStart(2, '0')}:${String(minutes).padStart(2, '0')}`;
      }
      formattedDate += ` ${timeStr}`;
    }
    
    return formattedDate;
  } catch (error) {
    console.error('Error formatting datetime for display:', error);
    return dateInput.toString();
  }
};

/**
 * Convert DD/MM/YYYY format to YYYY-MM-DD for HTML date input
 * @param {string} ddmmyyyy - Date in DD/MM/YYYY format
 * @returns {string} - Date in YYYY-MM-DD format for HTML input
 */
export const convertDDMMYYYYToHTMLDate = (ddmmyyyy) => {
  if (!ddmmyyyy || !ddmmyyyy.includes('/')) {
    return '';
  }
  
  try {
    const [day, month, year] = ddmmyyyy.split('/');
    return `${year}-${month.padStart(2, '0')}-${day.padStart(2, '0')}`;
  } catch (error) {
    console.error('Error converting DD/MM/YYYY to HTML date:', error);
    return '';
  }
};

/**
 * Convert YYYY-MM-DD format to DD/MM/YYYY for display
 * @param {string} yyyymmdd - Date in YYYY-MM-DD format
 * @returns {string} - Date in DD/MM/YYYY format
 */
export const convertHTMLDateToDDMMYYYY = (yyyymmdd) => {
  if (!yyyymmdd || !yyyymmdd.includes('-')) {
    return '';
  }
  
  try {
    const [year, month, day] = yyyymmdd.split('-');
    return `${day}/${month}/${year}`;
  } catch (error) {
    console.error('Error converting HTML date to DD/MM/YYYY:', error);
    return '';
  }
};

/**
 * Format date for input field (converts from DD/MM/YYYY to YYYY-MM-DD for HTML date input)
 * @param {string} dateString - Date string in various formats
 * @returns {string} - Date in YYYY-MM-DD format for HTML input
 */
export const formatDateForInput = (dateString) => {
  if (!dateString) return '';
  
  try {
    // Handle GMT format dates
    if (dateString.includes('GMT')) {
      const date = new Date(dateString);
      return date.toISOString().split('T')[0]; // Returns YYYY-MM-DD
    }
    
    // Handle ISO format dates
    if (dateString.includes('T')) {
      return dateString.split('T')[0];
    }
    
    // Handle DD/MM/YYYY format
    if (dateString.match(/^\d{2}\/\d{2}\/\d{4}$/)) {
      return convertDDMMYYYYToHTMLDate(dateString);
    }
    
    // If it's already in YYYY-MM-DD format
    if (dateString.match(/^\d{4}-\d{2}-\d{2}$/)) {
      return dateString;
    }
    
    // Try to parse as a regular date and convert
    const date = new Date(dateString);
    if (!isNaN(date.getTime())) {
      return date.toISOString().split('T')[0];
    }
    
    return '';
  } catch (error) {
    console.error('Error formatting date for input:', error);
    return '';
  }
};

/**
 * Format date for storage (converts from DD/MM/YYYY to YYYY-MM-DD for MySQL)
 * @param {string} dateString - Date in DD/MM/YYYY format from input
 * @returns {string} - Date in YYYY-MM-DD format for MySQL storage
 */
export const formatDateForStorage = (dateString) => {
  if (!dateString) return '';
  
  try {
    // If it's already in YYYY-MM-DD format, return as is
    if (dateString.match(/^\d{4}-\d{2}-\d{2}$/)) {
      return dateString;
    }
    
    // If it's in DD/MM/YYYY format, convert to YYYY-MM-DD
    if (dateString.match(/^\d{2}\/\d{2}\/\d{4}$/)) {
      const [day, month, year] = dateString.split('/');
      return `${year}-${month.padStart(2, '0')}-${day.padStart(2, '0')}`;
    }
    
    // Try to parse as a regular date and convert to YYYY-MM-DD
    const date = new Date(dateString);
    if (!isNaN(date.getTime())) {
      return date.toISOString().split('T')[0];
    }
    
    return '';
  } catch (error) {
    console.error('Error formatting date for storage:', error);
    return '';
  }
};

/**
 * Validate DD/MM/YYYY date format
 * @param {string} dateString - Date string to validate
 * @returns {boolean} - True if valid DD/MM/YYYY format
 */
export const isValidDDMMYYYY = (dateString) => {
  if (!dateString || !dateString.match(/^\d{2}\/\d{2}\/\d{4}$/)) {
    return false;
  }
  
  try {
    const [day, month, year] = dateString.split('/').map(num => parseInt(num, 10));
    const date = new Date(year, month - 1, day);
    
    return date.getDate() === day && 
           date.getMonth() === month - 1 && 
           date.getFullYear() === year;
  } catch (error) {
    return false;
  }
};

/**
 * Get current date in DD/MM/YYYY format
 * @returns {string} - Current date in DD/MM/YYYY format
 */
export const getCurrentDateDDMMYYYY = () => {
  const now = new Date();
  const day = String(now.getDate()).padStart(2, '0');
  const month = String(now.getMonth() + 1).padStart(2, '0');
  const year = now.getFullYear();
  return `${day}/${month}/${year}`;
};

/**
 * Get minimum date for date inputs (today) in YYYY-MM-DD format
 * @returns {string} - Today's date in YYYY-MM-DD format
 */
export const getMinDateForInput = () => {
  return new Date().toISOString().split('T')[0];
};
